### PR TITLE
Refactor2402 wip

### PR DESCRIPTION
**Pull request by Thinkpiet**
For developers: https://github.com/HSU-HPC/MaMiCo/wiki/Index-Cheatsheet#indexing-refactoring-cheatsheet

### CODE REVIEW

**Comment by Thinkpiet** [comment]
All Done! @rubenhorn(?)

File: ls1
[the remark this answers is not in the document]
Comment (Thinkpiet) [reply ls1:1]: Yes I they are (at least the latest one is)

**Review by rubenhorn — CHANGES_REQUESTED** [review]
I would recommend to simply remove all occurances of `@param indexConversion`.

The code looks good.
I verified that it can be compiled and that the tests pass.

However, the couette scenario with the example configuration in `couette.xml.template` does not run:
```
$ ../build/couette 
Run CouetteScenario...
Initialization: 2ms`
Finish CouetteScenario::initSolvers() 
ERROR simplemd::moleculemappings::VelocityStoermerVerletMapping::handleMolecule: Position 0 is out of range
Position: -nan , 30.6511 , 7.55553, molecule: 3135
Velocity: -nan , 0.189084 , 1.20591, molecule: 3135
OldVelocity: -1.48454 , 0.172614 , 1.31139, molecule: 3135
Force: -nan , 4.89854 , -14.1636, old: -9.08906 , 1.68924 , -28.0298
Old position: 22.0203 , 30.6501 , 7.54967
```
It still works as expected on the master branch. (Tag `legacyIndexing`)

File: coupling/indexing/Operations.h [hunk 1]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <sstream>
```
[the remark this answers is not in the document]
Comment (amartyads) [reply coupling/indexing/Operations.h:3]: i put it in debug since it's used in the debug output

**Comment by LouieVoit** [comment]
I still get some issues when using MPI this time. Without domain decomposition (i.e. `<mpi-configuration number-of-processes="1 ; 1 ; 1" ></mpi-configuration>`), I get this error:
`ERROR coupling::SendReceiveBuffer::triggerSending(): Could not send from rank 0 to the same rank! No receive buffer available! Tag (in case of MPI; not used here): 500001`


When run with domain decomposition, with and without multi MD instances, I get this one:
`ERROR coupling::SendReceiveBuffer:triggerSending(): Send- and receive-buffer sizes do not match for information processed on same rank!
Recv-buffer size=378, send-buffer size=492`

**Comment by Thinkpiet** [comment]
TODOS 

- [ ] Introduce the abstract class BaseCellContainer and the class FlexibleCellContainer. CellContainer and FlexibleCellContainer are subblasses of BaseCellContainer.
- [ ] The iterator in BaseCellContainer returns a std::pair<CellIndex, CouplingCell*>. std::tie shall be used to unpack the pair returned by the iterator.
- [ ] All the files in the sendrecv folder should be refactored
- [ ] In CouetteScenario.h:
    - [ ] Remove the CouplingBuffer structure and replace the send and receive buffers and indices by two FlexibleCellContainer
    - [ ] Rename sendBufffer and recvBuffer to respectively macroToMDBuffer and mdToMacroBuffer

File: test/unit/coupling/datastructures/FlexibleCellContainerTest.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  void test() {
+    using namespace coupling::datastructures;
+    std::vector<I00*> idxs;
+    std::vector<CouplingCell<3>*> couplingCells;
+    for (auto idx : I00()) {
+      // idxs.push_back(&(*idx));
+      std::cout << idx << std::endl;
+      // couplingCells.push_back(new CouplingCell<3>());
+    }
+    // FlexibleCellContainer<3> container{couplingCells, idxs};
```
[the remark this answers is not in the document]
Comment (LouieVoit) [reply test/unit/coupling/datastructures/FlexibleCellContainerTest.cpp:40]: Yes. I will update the test. It's just temporary.

**Comment by rubenhorn** [comment]
At [7dfabc0](https://github.com/HSU-HPC/MaMiCo/pull/45/commits/7dfabc094dfe9013a7a7c40336b1b649c4d0a447), the indexes being communicated (with 4 MPI ranks) match between [master](https://github.com/HSU-HPC/MaMiCo/pull/57) and the [refactor](https://github.com/HSU-HPC/MaMiCo/pull/59).

File: coupling/services/MultiMDCellService.h [hunk 3]
```diff
@@ -416,55 +423,43 @@ template <class LinkedCell, unsigned int dim> class coupling::services::MultiMDC
...
-     * If this is the first coupling step, we must init the post multi instance
-     * filter pipeline operating on averaged cell data. If you wish to not use post multi-instance filtering
-     * in deployment, you can simply leave the corresponding XML-Subtag empty.
-     * The ENABLE_POST_MULTI_INSTANCE_FILTERING flag is used for debugging purposes and shall be removed later.
-     */
     if (_postMultiInstanceFilterPipeline == nullptr) {
       // Init filter pipeline
-      _postMultiInstanceFilterPipeline = new coupling::filtering::FilterPipeline<dim>(_macroscopicCells, coupling::filtering::Scope::postMultiInstance,
-                                                                                      _multiMDService, _filterPipelineConfiguration.c_str());
+      // FIXME Invalid cell container
```
Comment (rubenhorn) [inline coupling/services/MultiMDCellService.h:445]: See #30

**Review by rubenhorn — APPROVED** [review]
LGTM

**Pull request merged**